### PR TITLE
[AdherentMessage] Refacto referent message 

### PR DESCRIPTION
**Pull request by ottaviano**

### CODE REVIEW

File: src/Mailchimp/Driver.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    public function getCampaignContent(string $campaignId): string
+    {
+        $response = $this->send('GET', sprintf('/campaigns/%s/content', $campaignId));
+
+        if ($this->isSuccessfulResponse($response)) {
+            return $this->toArray($response)['html'] ?? '';
+        }
+
+        $this->logger->warning(sprintf('[API] Error: %s', $response->getBody()), ['campaignId' => $campaignId]);
```
Comment (thomasbisignani) [inline src/Mailchimp/Driver.php]: We can use `$this->logger->error()` here no ?

File: src/Mailchimp/Driver.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    private function send(string $method, string $uri, array $body = []): ?ResponseInterface
+    {
+        try {
+            return $this->client->request(
+                $method,
+                '/3.0'.$uri,
+                ($body && \in_array($method, ['POST', 'PUT', 'PATCH'], true) ? ['json' => $body] : [])
+            );
+        } catch (RequestException $e) {
+            $this->logger->warning(sprintf(
```
Comment (thomasbisignani) [inline src/Mailchimp/Driver.php]: Same here to use `$this->logger->error()`

File: src/Mailchimp/Manager.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    public function deleteCampaign(string $campaignId): void
+    {
+        if (!$this->driver->deleteCampaign($campaignId)) {
+            $this->logger->warning(sprintf('Campaign "%s" was not be deleted', $campaignId));
```
Comment (thomasbisignani) [inline src/Mailchimp/Manager.php]: `Campaign "%s" has not been deleted`

File: src/Repository/AdherentMessageRepository.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+class AdherentMessageRepository extends ServiceEntityRepository
+{
+    use UuidEntityRepositoryTrait;
+
+    public function __construct(ManagerRegistry $registry)
+    {
+        parent::__construct($registry, AbstractAdherentMessage::class);
+    }
+
+    public function findAllForAuthor(Adherent $adherent): array
```
Comment (thomasbisignani) [inline src/Repository/AdherentMessageRepository.php]: `findAllByAuthor`

File: src/Repository/AdherentMessageRepository.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        return $this->findBy(['author' => $adherent]);
+    }
+
+    public function markAsSynchronized(AdherentMessageInterface $message): void
+    {
+        $message->setSynchronized(true);
+
+        $this->getEntityManager()->flush();
```
Comment (thomasbisignani) [inline src/Repository/AdherentMessageRepository.php]: IMO it's not the responsability of a repository to flush, what do you think to call the `flush()` in `AdherentMessageChangeCommandHandler` ?

File: src/Security/Voter/AuthoredVoter.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+{
+    protected function doVoteOnAttribute(string $attribute, Adherent $adherent, $subject): bool
+    {
+        /** @var AuthoredInterface $subject */
+        return $subject->getAuthor()->equals($adherent);
+    }
+
+    protected function supports($attribute, $subject)
+    {
+        return 'IS_AUTHOR_OF' === $attribute && $subject instanceof AuthoredInterface;
```
Comment (thomasbisignani) [inline src/Security/Voter/AuthoredVoter.php]: What about creating an `AuthorPermissions` with an `IS_AUTHOR_OF` const ?

File: src/Security/Voter/AuthoredVoter.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace AppBundle\Security\Voter;
+
+use AppBundle\Entity\Adherent;
+use AppBundle\Entity\AuthoredInterface;
+
+class AuthoredVoter extends AbstractAdherentVoter
```
Comment (thomasbisignani) [inline src/Security/Voter/AuthoredVoter.php]: WDYT to rename for `AuthorVoter` ?

File: translations/messages.fr.yml [hunk 8]
```diff
@@ -498,3 +498,14 @@ payment.message.holder_number_invalid: Votre numéro de carte semble être erron
 payment.message.card_issuer_invalid: Votre banque semble ne pas être reconnue par le système. Nous vous invitons à la contacter pour obtenir plus d'informations.
 payment.message.expiry_data_passed: La date de validité de votre carte semble être dépassée. Nous vous invitons à retenter l'opération avec un autre moyen de paiement.
 payment.message.card_absent: Votre carte bancaire semble ne pas être reconnue par le système. Nous vous invitons à contacter votre banque pour obtenir plus d'informations.
+
+# Adherent message
+adherent_message.created_successfully: 'Votre message a bien été créé.'
+adherent_message.updated_successfully: 'Votre message a bien été mis à jour.'
+adherent_message.deleted_successfully: 'Votre message a bien été supprimé'
```
Comment (thomasbisignani) [inline translations/messages.fr.yml]: missing `.` at the end

File: translations/messages.fr.yml [hunk 9]
```diff
@@ -498,3 +498,14 @@ payment.message.holder_number_invalid: Votre numéro de carte semble être erron
...
+adherent_message.filter_updated: 'Les filtres ont bien été sauvegardé'
```
Comment (thomasbisignani) [inline translations/messages.fr.yml]: `Les filtres ont bien été sauvegardés.`

File: src/Mailchimp/Manager.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    public function editCampaign(AdherentMessageInterface $message): bool
+    {
+        $requestBuilder = $this->requestBuildersLocator->get(CampaignRequestBuilder::class);
+
+        $editCampaignRequest = $requestBuilder->createEditCampaignRequestFromMessage($message);
+
+        if (!$campaignId = $message->getExternalId()) {
+            if (!$campaignId = $this->driver->createCampaign($editCampaignRequest)) {
+                throw new \RuntimeException(sprintf('Campaign for the message "%s" was not created', $message->getUuid()));
```
Comment (thomasbisignani) [inline src/Mailchimp/Manager.php]: `Campaign for the message "%s" has not been created`

File: src/Mailchimp/Manager.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    public function editCampaignContent(AdherentMessageInterface $message): bool
+    {
+        if (!$message->getExternalId()) {
+            throw new InvalidCampaignIdException(sprintf('Message "%s" does not have a valid campaign id', $message->getUuid()->toString()));
+        }
+
+        $requestBuilder = $this->requestBuildersLocator->get(CampaignRequestBuilder::class);
+
+        if (!$this->driver->editCampaignContent($message->getExternalId(), $requestBuilder->createContentRequest($message))) {
+            $this->logger->warning(sprintf('Campaign content of "%s" message was not modified', $message->getUuid()->toString()));
```
Comment (thomasbisignani) [inline src/Mailchimp/Manager.php]: Same here : `Campaign content of "%s" message has not been modified`

File: src/Mailchimp/Manager.php [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        if ($result) {
+            $this->logger->info(sprintf('Mailchimp member "%s" has been edited', $adherent->getUuidAsString()));
+
+            // Active/Inactive member's tags
+            $result = $this->driver->updateMemberTags(
+                $requestBuilder->createMemberTagsRequest($message->getEmailAddress(), $message->getRemovedTags())
+            );
+
+            if ($result) {
+                $this->logger->info(sprintf('Mailchimp member "%s" tags have been updated', $adherent->getUuidAsString()));
```
Comment (thomasbisignani) [inline src/Mailchimp/Manager.php]: `has been updated`

File: src/AdherentMessage/Handler/AdherentMessageChangeCommandHandler.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        $message = $this->repository->findOneByUuid($command->getUuid()->toString());
+
+        if (!$message || $message->isSynchronized()) {
+            return;
+        }
+
+        $needPersistCampaignId = null === $message->getExternalId();
+
+        if ($this->mailchimpManager->editCampaign($message)) {
+            if ($needPersistCampaignId) {
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/AdherentMessage/Handler/AdherentMessageChangeCommandHandler.php]: this call change the external id of $message `$this->mailchimpManager->editCampaign($message)`

File: src/Mailchimp/Manager.php (hunk 12 above)
Comment (ottaviano) [reply src/Mailchimp/Manager.php]: `they (tags) have been updated`

File: src/Controller/EnMarche/MessageController.php [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+     */
+    public function createMessageAction(string $prefix, Request $request, UserInterface $adherent, ObjectManager $manager): Response
+    {
+        $form = $this
+            ->createForm(AdherentMessageType::class)
+            ->handleRequest($request)
+        ;
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            $message = Factory::create($adherent, $form->getData());
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+
+    /**
+     * @Route("/{uuid}/filtrer", name="app_message_filter", methods={"GET"})
+     * @Security("is_granted('IS_AUTHOR_OF', message)")
+     */
+    public function filterMessageAction(string $prefix, Request $request, AbstractAdherentMessage $message, ObjectManager $manager): Response
+    {
+        // Reset Filter object
+        if ($request->query->has('reset') && $message->getFilter()) {
+            $message->setFilter(null);
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Mailchimp/Campaign/MailchimpObjectIdMapping.php [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    private $folderIds;
+    private $templateIds;
+
+    public function __construct(array $folderIds = [], array $templateIds = [])
+    {
+        $this->folderIds = $folderIds;
+        $this->templateIds = $templateIds;
+    }
+
+    public function getFolderIdForType(string $messageType): ?string
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Mailchimp/Campaign/MailchimpObjectIdMapping.php]: done

File: src/Mailchimp/Campaign/MailchimpObjectIdMapping.php [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        return $this->folderIds[$messageType] ?? null;
+    }
+
+    public function getTemplateIdForType(string $messageType): ?int
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Mailchimp/Campaign/MailchimpObjectIdMapping.php]: done

File: src/Mailchimp/Manager.php (hunk 3 above)
Comment (ottaviano) [reply src/Mailchimp/Manager.php]: done

File: src/Repository/AdherentMessageRepository.php (hunk 4 above)
Comment (ottaviano) [reply src/Repository/AdherentMessageRepository.php]: done

File: src/Repository/AdherentMessageRepository.php (hunk 5 above)
Comment (ottaviano) [reply src/Repository/AdherentMessageRepository.php]: done

File: src/Security/Voter/AuthoredVoter.php (hunk 6 above)
Comment (ottaviano) [reply src/Security/Voter/AuthoredVoter.php]: I have added `AuthorVoter::PERMISSION`, is it ok for you?

File: src/Security/Voter/AuthoredVoter.php (hunk 7 above)
Comment (ottaviano) [reply src/Security/Voter/AuthoredVoter.php]: done

File: translations/messages.fr.yml (hunk 8 above)
Comment (ottaviano) [reply translations/messages.fr.yml]: done thanks

File: translations/messages.fr.yml (hunk 9 above)
Comment (ottaviano) [reply translations/messages.fr.yml]: done

File: src/Mailchimp/Manager.php (hunk 10 above)
Comment (ottaviano) [reply src/Mailchimp/Manager.php]: done

File: src/Mailchimp/Manager.php (hunk 11 above)
Comment (ottaviano) [reply src/Mailchimp/Manager.php]: done

File: src/Security/Voter/AuthoredVoter.php (hunk 6 above)
Comment (thomasbisignani) [reply src/Security/Voter/AuthoredVoter.php]: Okay for me :ok_hand:

File: src/Controller/EnMarche/MessageController.php [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+ * @Route("/{prefix}messagerie", defaults={"prefix": ""}, requirements={"prefix": "|espace-referent/"})
+ *
+ * @Security("is_granted('ROLE_ADHERENT_MESSAGE')")
+ */
+class MessageController extends AbstractController
+{
+    /**
+     * @Route(name="app_message_list", methods={"GET"})
+     */
+    public function messageListAction(string $prefix, UserInterface $adherent, AdherentMessageRepository $repository): Response
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+            ['messages' => $repository->findAllByAuthor($adherent)]
+        );
+    }
+
+    /**
+     * @Route("/creer", name="app_message_create", methods={"GET", "POST"})
+     *
+     * @param Adherent|UserInterface $adherent
+     */
+    public function createMessageAction(string $prefix, Request $request, UserInterface $adherent, ObjectManager $manager): Response
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        if ($form->isSubmitted() && $form->isValid()) {
+            $message = AdherentMessageFactory::create($adherent, $form->getData());
+
+            $manager->persist($message);
+            $manager->flush();
+
+            $this->addFlash('info', 'adherent_message.created_successfully');
+
+            if ($form->get('next')->isClicked()) {
+                return $this->redirectToRoute('app_message_filter', ['prefix' => $prefix, 'uuid' => $message->getUuid()->toString()]);
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+        return $this->renderTemplate($prefix, 'message/create.html.twig', ['form' => $form->createView()]);
+    }
+
+    /**
+     * @Route("/{uuid}/modifier", name="app_message_update", methods={"GET", "POST"})
+     * @Security("is_granted('IS_AUTHOR_OF', message)")
+     */
+    public function updateMessageAction(string $prefix, Request $request, AbstractAdherentMessage $message, ObjectManager $manager): Response
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            $message->updateFromDataObject($dataObject);
+
+            $manager->flush();
+
+            $this->addFlash('info', 'adherent_message.updated_successfully');
+
+            if ($form->get('next')->isClicked()) {
+                return $this->redirectToRoute('app_message_filter', ['prefix' => $prefix, 'uuid' => $message->getUuid()->toString()]);
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+    public function getMessageTemplateAction(AbstractAdherentMessage $message, Manager $manager): Response
+    {
+        return new Response($manager->getCampaignContent($message));
+    }
+
+    /**
+     * @Route("/{uuid}/filtrer", name="app_message_filter", methods={"GET"})
+     * @Security("is_granted('IS_AUTHOR_OF', message)")
+     */
+    public function filterMessageAction(string $prefix, Request $request, AbstractAdherentMessage $message, ObjectManager $manager): Response
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+    {
+        // Reset Filter object
+        if ($request->query->has('reset') && $message->getFilter()) {
+            $message->resetFilter();
+            $manager->flush();
+
+            return $this->redirectToRoute('app_message_filter', ['prefix' => $prefix, 'uuid' => $message->getUuid()->toString()]);
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+        // Get Message filter or create a new object
+        $filter = ($message->getFilter() ?? FilterFactory::create($message->getType()))->handleRequest($request);
+
+        /** @var ReferentFilterDataObject $filter */
+        if ($filter->hasToken()) {
+            // Redirect if CSRF is invalid
+            if (!$this->isCsrfTokenValid(ReferentController::TOKEN_ID, $filter->getToken())) {
+                return $this->redirectToRoute('app_message_filter', ['prefix' => $prefix, 'uuid' => $message->getUuid()->toString()]);
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/Controller/EnMarche/MessageController.php [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+            $message->setFilter(clone $filter);
+            $manager->flush();
+
+            $this->addFlash('info', 'adherent_message.filter_updated');
+
+            return $this->redirectToRoute('app_message_list', ['prefix' => $prefix, 'uuid' => $message->getUuid()->toString()]);
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MessageController.php]: multiline

File: src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+    public function postPersist(LifecycleEventArgs $args): void
+    {
+        $object = $args->getObject();
+
+        if (!$object instanceof AdherentMessageInterface) {
+            return;
+        }
+
+        $this->dispatchMessage($object);
+    }
```
Comment (Raulnet) [inline src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php]:
```if ($object instanceof AdherentMessageInterface) {
        $this->dispatchMessage($object);
        }
         ```
Bettter ?

File: src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+    public function postUpdate(LifecycleEventArgs $args): void
+    {
+        $object = $args->getObject();
+
+        if (!$object instanceof AdherentMessageInterface || $object->isSynchronized()) {
+            return;
+        }
+
+        $this->dispatchMessage($args->getObject());
+    }
```
Comment (Raulnet) [inline src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php]: if ($object instanceof AdherentMessageInterface && !$object->isSynchronized()) {
          $this->dispatchMessage($args->getObject());
        }

File: templates/message/filter/referent.html.twig [hunk 29]
```diff
@@ -0,0 +1,177 @@
...
+                    <div class="form l__row l__row--wrap">
+                        {% for key, value in adherent_interests %}
+                            <div class="form__checkbox">
+                                <input type="checkbox" id="referent_pin_interests_{{ loop.index0 }}" name="{{ constant('PARAMETER_INTEREST', filter) }}[]" value="{{ key }}" {{ key in filter.queryInterests ? 'checked' }}>
+                                <label  for="referent_pin_interests_{{ loop.index0 }}">{{ value }}</label>
+                            </div>
+                        {% endfor %}
+                    </div>
+                    <legend class="text--gray text--small">
+                        Exemple : Si vous cochez 2 intérêts, vous filtrerez les adhérents qui suivent au moins ces 2 intérêts
```
Comment (JeremieSamson) [inline templates/message/filter/referent.html.twig]:
```suggestion
                        Exemple : Si vous cochez 2 intérêts, vous filtrerez les adhérents qui suivent au moins sur ces 2 intérêts
```

File: src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php (hunk 27 above)
Comment (ottaviano) [reply src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php]: agree, changed 👍

File: src/Controller/EnMarche/MessageController.php (hunk 18 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 19 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 20 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 21 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 22 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 23 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 24 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 25 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/Controller/EnMarche/MessageController.php (hunk 26 above)
Comment (ottaviano) [reply src/Controller/EnMarche/MessageController.php]: done

File: src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php (hunk 28 above)
Comment (ottaviano) [reply src/AdherentMessage/Listener/AdherentMessageChangeSubscriber.php]: changed thanks

File: templates/message/filter/referent.html.twig (hunk 29 above)
Comment (ottaviano) [reply templates/message/filter/referent.html.twig]: `adherents suivent quelque chose` vs `adherents suivent quelque chose`

File: src/Entity/AuthoredInterface.php [hunk 30]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace AppBundle\Entity;
+
+interface AuthoredInterface
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Entity/AuthoredInterface.php]: @JeremieSamson, I have chosen `Authored` like existing trait `AuthoredTrait`
https://github.com/EnMarche/en-marche.fr/blob/master/src/Entity/AuthoredTrait.php#L5

**Pull request merged**